std/experimental/typecons.d

```diff
@@ -0,0 +1,731 @@
+// Written in the D programming language.
+
+/**
+This module implements a variety of type constructors, i.e., templates
+that allow construction of new, useful general-purpose types.
+
+Source:    $(PHOBOSSRC std/_typecons.d)
+
+Macros:
+
+WIKI = Phobos/StdVariant
+
+Synopsis:
+
+----
+// value tuples
+alias Coord = Tuple!(float, "x", float, "y", float, "z");
+Coord c;
+c[1] = 1;       // access by index
+c.z = 1;        // access by given name
+alias DicEntry = Tuple!(string, string); // names can be omitted
+
+// Rebindable references to const and immutable objects
+void bar()
+{
+    const w1 = new Widget, w2 = new Widget;
+    w1.foo();
+    // w1 = w2 would not work; can't rebind const object
+    auto r = Rebindable!(const Widget)(w1);
+    // invoke method as if r were a Widget object
+    r.foo();
+    // rebind r to refer to another object
+    r = w2;
+}
+----
+
+Copyright: Copyright the respective authors, 2008-
+License:   $(WEB boost.org/LICENSE_1_0.txt, Boost License 1.0).
+Authors:   $(WEB erdani.org, Andrei Alexandrescu),
+           $(WEB bartoszmilewski.wordpress.com, Bartosz Milewski),
+           Don Clugston,
+           Shin Fujishiro,
+           Kenji Hara
+ */
+module std.experimental.typecons;
+
+import std.meta; // : AliasSeq, allSatisfy;
+import std.traits;
+
+import std.typecons: Tuple, tuple, Bind, DerivedFunctionType,
+       isImplicitlyConvertible, mixinAll, staticIota,
+       GetOverloadedMethods;
+
+private
+{
+    pragma(mangle, "_d_toObject")
+    extern(C) pure nothrow Object typecons_d_toObject(void* p);
+}
+
+/*
+ * Avoids opCast operator overloading.
```

Comment:
This is part of the original wrap, I'm not making changes to this functionality especially with this case where I don't know what it would break.